[WIP] build.cake Refactor Semantic Version

devlead opened the pull request:
Currently semver is used in a couple of places before it's assigned (i.e. build output en setup message), also on posix it was never assigned.

This WIP PR proposes the following changes
* Semantic version now calculated at setup
* SolutionInfo used as fallback
* Directory paths instantiated post semver
* Use V3 nuget feeds
* Remove code duplication (create nuget & copy files)

gep13 commented on build.cake:
```diff
@@ -42,7 +43,63 @@ var password = EnvironmentVariable("CAKE_GITHUB_PASSWORD");
+
+            // Due to the way that GitVersion is executed on AppVeyor, the Environment Variables although populated
+            // are not accessible yet, so have to run GitVersion again, using OutputType of JSON
+            if(string.IsNullOrEmpty(semVersion))
```
The contents of this if statement is essentially duplicated in the else statement below.  So I think we could remove this duplication, and just always run this code, just make sure to run the above code, if running on AppVeyor.

devlead replied:
@gep13 true, I'll look at that.

gep13 commented on build.cake:
```diff
@@ -42,7 +43,63 @@ var password = EnvironmentVariable("CAKE_GITHUB_PASSWORD");
+    if (isRunningOnWindows)
+    {
+        Information("Calculating Semantic Version");
+        if (AppVeyor.IsRunningOnAppVeyor)
```
This might not be enough.  Given that we are running on multiple build servers, i.e. Jenkins, TeamCity, Bamboo, etc, we want GitVersion to be ran with `/output buildserver` so that it can do it's branch normalisation, where required.  Perhaps what we need here is a check to see if we are **not** running locally.

devlead replied:
@gep13 true, I'll look at that.

devlead commented:
@gep13 so I went a little crazy... but it still compiles :wink: Still WIP, but I think ready for feedback.

gep13 commented:
@devlead ha ha, you did go a little crazy! I will need to take a closer look at this, and get back to you.

For now, can you rebase against develop and re-push?

patriksvensson commented on build/BuildPackages.cake:
```diff
@@ -0,0 +1,35 @@
+public class BuildPackages
```
Should add a `#load "./BuildPackage.cake"` here. It's not needed but it will make it explicit that this is needed.

patriksvensson commented on build/BuildPaths.cake:
```diff
@@ -0,0 +1,94 @@
+        }
+
+        var buildDir = context.Directory("./src/Cake/bin") + context.Directory(configuration);
+        DirectoryPath artifactsDir = context.Directory("./artifacts") + context.Directory("v" + semVersion);
```
Should be declared as `var`.

patriksvensson commented on build/BuildFiles.cake:
```diff
@@ -0,0 +1,22 @@
+public class BuildFiles
+{
+    public ICollection<FilePath> CakeAsseemblyPaths { get; private set; }
```
Should be called `CakeAssemblyPaths` (spelling).

patriksvensson commented on build/BuildFiles.cake:
```diff
@@ -0,0 +1,22 @@
+    public ICollection<FilePath> RepoFilesPaths { get; private set; }
+    public ICollection<FilePath> ArtifactsSourcePaths { get; private set;}
+    public FilePath ZipArtifactPath { get;  private set; }
+    public BuildFiles(
```
New line

patriksvensson commented on build/BuildPackages.cake:
```diff
@@ -0,0 +1,35 @@
+    public ICollection<BuildPackage> Chocolatey { get; private set;}
+
+    public static BuildPackages GetPackages(
+
```
Empty line

patriksvensson commented on build/BuildPackages.cake:
```diff
@@ -0,0 +1,35 @@
+
+    private static Func<string, BuildPackage> BuildPackage(DirectoryPath nugetRooPath, string semVersion, bool isChocolateyPackage = false)
+    {
+        return package=> new BuildPackage(
```
Space before `=>`

patriksvensson commented on build/BuildPackages.cake:
```diff
@@ -0,0 +1,35 @@
+    private static Func<string, BuildPackage> BuildPackage(DirectoryPath nugetRooPath, string semVersion, bool isChocolateyPackage = false)
+    {
+        return package=> new BuildPackage(
+                        id:package,
```
Space after `:`

patriksvensson commented on build/BuildPackages.cake:
```diff
@@ -0,0 +1,35 @@
+    {
+        return package=> new BuildPackage(
+                        id:package,
+                        nuspecPath:string.Concat("./nuspec/", package, ".nuspec"),
```
Space after `:`

patriksvensson commented on build/BuildPackages.cake:
```diff
@@ -0,0 +1,35 @@
+        return package=> new BuildPackage(
+                        id:package,
+                        nuspecPath:string.Concat("./nuspec/", package, ".nuspec"),
+                        packagePath:nugetRooPath.CombineWithFilePath(string.Concat(package, ".", semVersion, ".nupkg")),
```
Space after `:`

patriksvensson commented on build/BuildPackages.cake:
```diff
@@ -0,0 +1,35 @@
+                        id:package,
+                        nuspecPath:string.Concat("./nuspec/", package, ".nuspec"),
+                        packagePath:nugetRooPath.CombineWithFilePath(string.Concat(package, ".", semVersion, ".nupkg")),
+                        isChocolateyPackage:isChocolateyPackage
```
Space after `:`

patriksvensson commented on build/BuildDirectories.cake:
```diff
@@ -0,0 +1,26 @@
+    public DirectoryPath NugetRoot { get; private set; }
+    public DirectoryPath ArtifactsBin { get; private set;}
+    public ICollection<DirectoryPath> ToClean { get; private set;}
+    public BuildDirectories (
```
Should be an empty line here.

patriksvensson commented on build/BuildPackages.cake:
```diff
@@ -0,0 +1,35 @@
+        )
+    {
+        var toNugetPackage = BuildPackage(nugetRooPath, semVersion);
+        var toChocolateyPackage = BuildPackage(nugetRooPath, semVersion, isChocolateyPackage:true);
```
Space after `:`

patriksvensson commented on build/parameters.cake:
```diff
@@ -0,0 +1,74 @@
+#load "./paths.cake"
+#load "./packages.cake"
+#load "./version.cake"
+public class BuildParameters
```
Sorry, but new line here :smile:

gep13 commented on build/version.cake:
```diff
@@ -0,0 +1,71 @@
+
+        if (context.IsRunningOnWindows())
+        {
+            context.Information("Calculating Semantic Version");
```
@devlead minor point, but when writing messages to output, if you are "away" to do something, I prefer to see this as "Calculating Semantic Version..." The ... indicating that it is away to start. Thoughts?

patriksvensson replied:
@gep13 @devlead Agree with Gary's comment.

devlead replied:
@gep13 I have no objection against that, will correct.

gep13 commented on build/version.cake:
```diff
@@ -0,0 +1,71 @@
+                milestone = string.Concat("v", version);
+            }
+
+            GitVersion assertedVersions = context.GitVersion(new GitVersionSettings
```
@devlead I think we need a check here as well.  Running GitVersion again will only be required **if** we are unable to access the environment variables.

devlead replied:
@gep13 ok, won't hurt.

gep13 commented on build/version.cake:
```diff
@@ -0,0 +1,71 @@
+
+        if (string.IsNullOrEmpty(version) || string.IsNullOrEmpty(semVersion))
+        {
+            context.Information("Fetching verson from SolutionInfo");
```
@devlead same again here with ...

devlead replied:
@gep13 sure :+1:

devlead replied:
Also correcting the typo ;)

gep13 commented on build/version.cake:
```diff
@@ -0,0 +1,71 @@
+    public string Milestone { get; private set; }
+    public string CakeVersion { get; private set; }
+
+    public static BuildVersion CalculatingSemanticVersion(
```
@devlead should this be `CalculateSemanticVersion`?

devlead replied:
@gep13 yes

gep13 commented on build.cake:
```diff
@@ -42,7 +12,37 @@ var password = EnvironmentVariable("CAKE_GITHUB_PASSWORD");
+        BuildPackages.GetPackages(
+        nugetRooPath: parameters.Paths.Directories.NugetRoot,
+        semVersion: parameters.Version.SemVersion,
+        packageIds: new [] { "Cake", "Cake.Core", "Cake.Common", "Cake.Testing", "Cake.NuGet" },
```
@devlead to be consistent, should this be `nugetPackageIds`?

devlead replied:
@gep13 yes

gep13 commented on build.cake:
```diff
@@ -42,7 +12,37 @@ var password = EnvironmentVariable("CAKE_GITHUB_PASSWORD");
+        )
+    );
+
+    Information("Building version {0} of Cake ({1}, {2}) using version {3} of Cake.",
```
@devlead ...

devlead replied:
@gep13 ...yes :wink:

gep13 commented on build.cake:
```diff
@@ -50,11 +50,7 @@ Setup(() =>
-    CleanDirectories(new DirectoryPath[] {
-        buildResultDir, binDir, testResultsDir, nugetRoot});
-});
+    .Does(() => CleanDirectories(parameters.Paths.Directories.ToClean));
```
@devlead hmm, this looks _wrong_ to me.  I would prefer to still include the {}'s, but that might just be a personal preference.

devlead replied:
Just trying to optimize my keystrokes before they run out :smile:  But I'll adjust, as I'm good with both :wink:

gep13 commented on build.cake:
```diff
@@ -160,137 +105,78 @@ Task("Copy-Files")
-
-    var files = GetFiles(binDir.Path.FullPath + "/*")
-      - GetFiles(binDir.Path.FullPath + "/*.Testing.*");
+    var files = GetFiles( parameters.Paths.Directories.ArtifactsBin.FullPath + "/*")
```
@devlead extra space after opening (

gep13 commented on build.cake:
```diff
@@ -160,137 +105,78 @@ Task("Copy-Files")
-    .IsDependentOn("Run-GitVersion")
     .IsDependentOn("Package")
-    .WithCriteria(() => isRunningOnWindows)
+    .WithCriteria(() => parameters.IsRunningOnWindows)
```
@devlead technically, chocolatey can run on Mac/Linux using mono, but I think that this makes sense.

devlead replied:
yeah, leave it as is.

gep13 commented on build.cake:
```diff
@@ -160,137 +105,78 @@ Task("Copy-Files")
+    }
 });
 
 Task("Update-AppVeyor-Build-Number")
```
@devlead this task isn't required.  Running GitVersion with `/output buildserver` already takes care of that.

devlead replied:
Ok! Didn't know I'll remove it then.

gep13 commented:
@devlead based on the other conversations that we have had around this, do we want to change the top level Publish task, so that the contained Tasks are done, automagically, when running on AppVeyor, and when it is the main branch that is being built?  Additionally, we might want to include the presence of a Tag?  That would mean we would need to change the calls to GitReleaseManager a little bit though.

devlead commented:
@gep13 we could publish automagically on AppVeyor, but perhaps that should be step 2, refactored allot as is :wink: 
Think I've addressed most feedback given, ping me otherwise.
If we're happy with it and want to try it I could reopen the PR against master and we could to a _"hotfix"_ release to test our release process.

gep13 commented:
>@devlead said...
If we're happy with it and want to try it I could reopen the PR against master and we could to a "hotfix" release to test our release process.

Yes, I think that would be a good idea.  @patriksvennson what are your thoughts?

devlead commented:
Replaced by #768
